components/camera/vcamera/vcamera.go

```diff
@@ -0,0 +1,210 @@
+	err                     error
+	cancelCtx               context.Context
+	cancelFn                func()
+	activeBackgroundWorkers sync.WaitGroup
+}
+
+// Builder creates a new vcamera.Config builder object.
+func Builder() *Config {
+	cancelCtx, cancelFn := context.WithCancel(context.Background())
+	return &Config{
+		devices:   make(map[int]bool),
+		cancelFn:  cancelFn,
+		cancelCtx: cancelCtx,
+	}
+}
+
+// NewCamera lazily creates a new camera. No cameras are actually created until Stream is called.
+func (c *Config) NewCamera(id int, label string, res Resolution) *Config {
+	if c.err != nil {
+		return c
+	}
+
+	if _, ok := c.devices[id]; ok {
+		c.err = fmt.Errorf("duplicate camera id %d", id)
+		return c
+	}
+
+	c.devices[id] = true
+	c.deviceKeys = append(c.deviceKeys, strconv.Itoa(id))
+	c.labels = append(c.labels, fmt.Sprintf("\"%s\"", label)) // add quotes
+	c.resolutions = append(c.resolutions, res)
+	return c
+}
+
+func createCameras(c *Config) error {
+	devices := fmt.Sprintf("%s=%s", "video_nr", strings.Join(c.deviceKeys, ","))
+	labels := fmt.Sprintf("%s=%s", "card_label", strings.Join(c.labels, ","))
+
+	cmd := fmt.Sprintf("sudo modprobe v4l2loopback %s %s", devices, labels)
+	out, err := exec.Command("bash", "-c", cmd).CombinedOutput() //nolint:gosec
+	if err != nil {
+		return errors.New(string(out))
+	}
+	return nil
+}
+
+func startStream(config *Config, idx int) (<-chan struct{}, error) {
+	//nolint:gosec
+	cmd := exec.CommandContext(config.cancelCtx, "bash", "-c", fmt.Sprintf(
+		"gst-launch-1.0 -v videotestsrc "+
+			"! video/x-raw,format=YUY2,width=320,height=240 "+
+			"! videoscale "+
+			"! video/x-raw,format=YUY2,width=%d,height=%d "+
+			"! v4l2sink device=/dev/video%s",
+		config.resolutions[idx].Width, config.resolutions[idx].Height, config.deviceKeys[idx]))
+
+	// capture output from stdout
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	// wait for cmd to finish to avoid leak
+	config.activeBackgroundWorkers.Add(1)
+	go func() {
+		defer config.activeBackgroundWorkers.Done()
+		cmd.Wait() //nolint:errcheck,gosec
+	}()
+
+	// broadcast channel, closed when stream is in state "PLAYING"
+	readyCh := make(chan struct{})
+	scanner := bufio.NewScanner(stdout)
+
+	config.activeBackgroundWorkers.Add(1)
+	go func() {
+		defer config.activeBackgroundWorkers.Done()
+		for {
+			select {
+			case <-config.cancelCtx.Done():
+				return
+			default:
+				scanner.Scan()
+				// looks for state "PLAYING" in stdout
+				if line := scanner.Text(); strings.Contains(line, "PLAYING") {
+					close(readyCh)
+					return
+				}
+			}
+		}
+	}()
+	return readyCh, nil
+}
+
+// Stream starts streaming videos to all virtual cameras.
+//
+// This function blocks until all virtual cameras successfully start streaming, or an error occurs.
+// Shutdown will stop all streams started by this method.
+func (c *Config) Stream() (*Config, error) {
```

Comment:
why does a Stream return a Config? I'd imagine a Stream would return some kind of stream that could be shutdown